enf/resource_enf_endpoint_ratelimit.go

```diff
@@ -0,0 +1,181 @@
+	d.Set("bytes_burst_size", endpointRateLimit.BytesBurstSize)
+	d.Set("inherit", endpointRateLimit.Inherit)
+	return nil
+}
+
+func endpointMaxRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	endpointRateLimit, resp, err := client.Endpoint.GetMaxRateLimits(context.Background(), d.Get("endpoint").(string))
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.SetId(d.Get("endpoint").(string))
+	d.Set("packets_per_second", endpointRateLimit.PacketsPerSecond)
+	d.Set("packets_burst_size", endpointRateLimit.PacketsBurstSize)
+	d.Set("bytes_per_second", endpointRateLimit.BytesPerSecond)
+	d.Set("bytes_burst_size", endpointRateLimit.BytesBurstSize)
+	d.Set("inherit", endpointRateLimit.Inherit)
+	return nil
+}
+func resourceEnfEndpointRateLimitUpdate(d *schema.ResourceData, meta interface{}) error {
+	if err := validateRateLimitRequest(d); err != nil {
+		return err
+	}
+
+	if d.Get("type") == "current" {
+		return endpointCurrentRateLimitUpdate(d, meta)
+	} else if d.Get("type") == "max" {
+		return endpointMaxRateLimitUpdate(d, meta)
+	} else {
+		return errors.New("The rate limit type must be either current or max")
+	}
+}
+
+func endpointCurrentRateLimitUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	endpointRateLimitRequest := &enf.EndpointRateLimits{
+		PacketsPerSecond: enf.Int(d.Get("packets_per_second").(int)),
+		PacketsBurstSize: enf.Int(d.Get("packets_burst_size").(int)),
+		BytesPerSecond:   enf.Int(d.Get("bytes_per_second").(int)),
+		BytesBurstSize:   enf.Int(d.Get("bytes_burst_size").(int)),
+		Inherit:          enf.Bool(d.Get("inherit").(bool)),
+	}
+
+	_, _, err := client.Endpoint.SetCurrentRateLimits(context.Background(), endpointRateLimitRequest, d.Get("endpoint").(string))
+	if err != nil {
+		return err
+	}
+
+	return resourceEnfEndpointRateLimitRead(d, meta)
+}
+
+func endpointMaxRateLimitUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	endpointRateLimitRequest := &enf.EndpointRateLimits{
+		PacketsPerSecond: enf.Int(d.Get("packets_per_second").(int)),
+		PacketsBurstSize: enf.Int(d.Get("packets_burst_size").(int)),
+		BytesPerSecond:   enf.Int(d.Get("bytes_per_second").(int)),
+		BytesBurstSize:   enf.Int(d.Get("bytes_burst_size").(int)),
+		Inherit:          enf.Bool(d.Get("inherit").(bool)),
+	}
+
+	_, _, err := client.Endpoint.SetMaxRateLimits(context.Background(), endpointRateLimitRequest, d.Get("endpoint").(string))
+	if err != nil {
+		return err
+	}
+
+	return resourceEnfEndpointRateLimitRead(d, meta)
+}
+
+func resourceEnfEndpointRateLimitDelete(d *schema.ResourceData, meta interface{}) error {
+	// Since we cannot "delete" a rate limit, we'll instead reset to the default values by updating the inherited flag to 'true'.
+	d.Set("inherit", true)
+
+	switch d.Get("type") {
```

Comment:
Can you just call `resourceEnfEndpointRateLimitUpdate(d, meta)`, rather than having to repeat the switch-case logic here?

E.g.,
```go
d.Set("inherit", true)
d.Set("packets_per_second", nil)
d.Set("packets_burst_size", nil)
d.Set("bytes_per_second", nil)
d.Set("bytes_burst_size", nil)

err := resourceEnfEndpointRateLimitUpdate(d, meta)
if err != nil {
    return err
}

return nil
```

I'm honestly not sure.

Reply:
We chatted a bit about this, but I don't want to call `resourceEnfEndpointRateLimitUpdate` because `validateRateLimitRequest` will be called. And since we simulate delete by setting `inherit=true`, the validation would fail because we have no way of telling Terraform to "unset" the other resource properties. 

I'll need to leave the delete methods as-is for now, despite the fact the switches make the code overly verbose.